src/current/v23.1/changefeed-messages.md

```diff
@@ -23,35 +23,137 @@ This page describes the format and behavior of changefeed messages. You will fin
+For [webhook sinks](changefeed-sinks.html#webhook-sink), the response format arrives as a batch of changefeed messages with a `payload` and `length`.
 
 ~~~
 {"payload": [{"after" : {"a" : 1, "b" : "a"}, "key": [1], "topic": "foo"}, {"after": {"a": 1, "b": "b"}, "key": [1], "topic": "foo" }], "length":2}
 ~~~
 
-See [changefeed files](create-changefeed.html#files) for more detail on the file naming format for {{ site.data.products.enterprise }} changefeeds.
+[Webhook message batching](changefeed-sinks.html#webhook-sink-configuration) is subject to the same key [ordering guarantee](#ordering-guarantees) as other sinks. Therefore, as messages are batched, you will not receive two batches at the same time with overlapping keys. You may receive a single batch containing multiple messages about one key, because ordering is maintained for a single key within its batch.
+
+Refer to [changefeed files](create-changefeed.html#files) for more detail on the file naming format for {{ site.data.products.enterprise }} changefeeds.
+
+## Message envelopes
+
+The _envelope_ defines the structure of a changefeed message. You can use the [`envelope`](create-changefeed.html#envelope) option to manipulate the changefeed envelope. The values that the `envelope` option accepts are compatible with different [changefeed sinks](changefeed-sinks.html), and the structure of the message will vary depending on the sink.
+
+{{site.data.alerts.callout_info}}
+Changefeeds created with [`EXPERIMENTAL CHANGEFEED FOR`](changefeed-for.html) or [`CREATE CHANGEFEED`](create-changefeed.html) with no sink specified (sinkless changefeeds) produce messages without the envelope metadata fields of changefeeds emitting to sinks.
+{{site.data.alerts.end}}
+
+The following sections provide examples of changefeed messages that are emitted when you specify each of the supported `envelope` options. Other [changefeed options](create-changefeed.html#options) can affect the message envelope and the what messages are emitted. Therefore, the examples are a guide for what you can expect when only the `envelope` option is specified.
+
+### `wrapped`
+
+`wrapped` is the default envelope structure for changefeed messages. This envelope contains an array of the primary key (or the key as part of the message metadata), a top-level field for the type of message, and the current state of the row (or `null` for [deleted rows](#delete-messages)).
+
+The message envelope contains a primary key array when your changefeed is emitting to a sink that does not have a message key as part of its protocol, e.g., cloud storage, webhook sinks, or Google Pub/Sub. By default, messages emitted to Kafka sinks do not have the primary key array, because the key is part of the message metadata. If you would like messages emitted to Kafka sinks to contain a primary key array, you can use the [`key_in_value`](create-changefeed.html#key-in-value) option. Refer to the following message outputs for examples of this.
```

Comment:
nit: can leave as is

```suggestion
The message envelope contains a primary key array when your changefeed is emitting to a sink that does not have a message key as part of its protocol, (e.g., cloud storage, webhook sinks, or Google Pub/Sub). By default, messages emitted to Kafka sinks do not have the primary key array, because the key is part of the message metadata. If you would like messages emitted to Kafka sinks to contain a primary key array, you can use the [`key_in_value`](create-changefeed.html#key-in-value) option. Refer to the following message outputs for examples on this.
```